tests/helpers/kubernetes/minikube.py

```diff
@@ -1,208 +1,326 @@
 import os
+import re
+import socket
 import tempfile
 import time
+import urllib
 from contextlib import contextmanager
 from functools import partial as p
 
 import docker
+import requests
 import semver
 import yaml
 from kubernetes import config as kube_config
 
-from tests.helpers.assertions import container_cmd_exit_0
+import tests.helpers.kubernetes.utils as k8s
+from tests.helpers.assertions import container_cmd_exit_0, tcp_socket_open
+from tests.helpers.formatting import print_dp_or_event
 from tests.helpers.kubernetes.agent import Agent
-from tests.helpers.kubernetes.utils import (
-    api_client_from_version,
-    container_is_running,
-    create_resource,
-    delete_resource,
-    get_all_logs,
-    get_free_port,
-    has_docker_image,
-    has_resource,
-    wait_for_deployment,
+from tests.helpers.util import (
+    container_ip,
+    fake_backend,
+    get_docker_client,
+    get_host_ip,
+    retry,
+    wait_for,
+    TEST_SERVICES_DIR,
 )
-from tests.helpers.util import container_ip, get_docker_client, wait_for
+from tests.packaging.common import get_container_file_content
 
+MINIKUBE_CONTAINER_NAME = "minikube"
+MINIKUBE_IMAGE_NAME = "minikube"
 MINIKUBE_VERSION = os.environ.get("MINIKUBE_VERSION")
-MINIKUBE_LOCALKUBE_VERSION = "v0.28.2"
-MINIKUBE_KUBEADM_VERSION = "v0.30.0"
-TEST_SERVICES_DIR = os.path.join(os.path.dirname(os.path.realpath(__file__)), "../../../test-services")
+MINIKUBE_LOCALKUBE_VERSION = "0.28.2"
+MINIKUBE_KUBEADM_VERSION = "0.34.1"
+MINIKUBE_IMAGE_TIMEOUT = int(os.environ.get("MINIKUBE_IMAGE_TIMEOUT", 300))
+MINIKUBE_KUBECONFIG_PATH = "/kubeconfig"
+K8S_API_PORT = 8443
+K8S_RELEASE_URL = "https://storage.googleapis.com/kubernetes-release/release/stable.txt"
+K8S_MIN_VERSION = "1.7.0"
+K8S_MIN_KUBEADM_VERSION = "1.11.0"
+
+
+def get_free_port():
```

Comment:
Note that this will actually bind the port and prevent it from being used without actually returning the socket that underlies it.

Reply:
Updated to close the socket when returning the port.